rfcs/005 - Unsurprising Functions and Scalability RFC.md

```diff
@@ -0,0 +1,215 @@
+
+### Final Proposal
+
+1. The simplest solution to the RPC behavior is to use [`pgbouncer`](https://www.pgbouncer.org/) to only allow one SQL statement per connection and fail out requests that do otherwise. This is a known configuration that every SQL editor we are aware of can handle, as well as several ORMs. This layer on top of the database should [allow us to enable proper SSL certificates](https://www.percona.com/blog/2021/06/28/enabling-ssl-tls-sessions-in-pgbouncer/) to improve trustworthiness with end users and eliminate (most) MITM possibilities.
+2. Efficiency gains as well as a reduction in complexity can be had by switching to the [`pgsql-http`](https://github.com/pramsey/pgsql-http) extension and having the RPC functions simply wait for the HTTP response from the engine, though they must *all* finish before the HTTP connection times out. This requires getting off of RDS, as this [is not one of the blessed extensions](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-extensions.html). The HTTP timeout issue can be avoided entirely if the engine is a sidecar process to the Postgres instance and they communicate over `localhost`.
+3. Take the final few HTTP endpoints exposed to the end users and dashboard in the engine and turn most of them into RPC functions in the `metadata` table, and close off HTTP access to the outer world entirely, exposing only the database. The `/run` endpoint would turn into a lambda function for the dashboard, instead. This lets us simplify our deployment artifact into a singular Docker container with all three (Postgres, Engine, pgbouncer) communicating directly to each other.
+4. Deploy this to a horizontally-scaling collection of EC2 instances with EBS volumes for the database, initially only one such instance, but scaling up to more requires a simple routing layer above that decides the target EC2 instance by the database ID being connected to, with the sharding on the number of logical databases per actual database being determined by the scaling factors we run into.
+
+This singular docker container can also be pushed to Dockerhub and make the deployments even simpler, being just downloading a new docker image, restarting the container, and pruning the dead image(s), with deployments taking seconds rather than minutes, though with a brief moment of downtime (before we implement the routing layer, which could hold external connections open until downstream is up again, then making it zero-downtime).
+
+It would also be possible to consider docker-container-per-user-database (with multiple of them per EC2 instance), which would allow us to drop maintaining multiple module versions at the same time within the engine. `iasql_upgrade` would become an *actual* upgrade of the engine to a newer version triggered by the end user, instead, but this and the routing layer could be done progressively in the future after the main transition in the four points above.
```

Comment:
I want to echo that like (4) and the previous sentence this would be follow-up work. 

I hadn't thought of this. However, to fully get rid of multiple module versions the user call to `upgrade` would also require a deploy to their respective container?

Reply:
Basically there would be no `iasql_upgrade` as a callable function anymore. We'd just make upgrading from earlier versions of IaSQL something it does during startup just like Postgres auto-upgrades its on-disk representation during startup if you point a new version of Postgres at an older version's data (only up to so many versions behind, though, which we would also preserve).

So the `upgrade` code wouldn't go away, but 1) it would be fully contained in a singular logical block instead of split between "old version upgrade RPC triggering" and "new version continuation function that needs to be compatible with all supported older versions" and 2) it would only run during startup before anything else is set up on the engine, so there's no weird issues of the upgrade running in parallel with an apply or install.

Upgrading in the dashboard would still exist, but it would call something else (perhaps a lambda to upgrade the database/docker in question)? And yes, this is definitely not required for the initial pass, but this particular architecture makes it possible to have per-user IaSQL DB versioning (like we currently support) but without all of the complexities to the codebase the current approach has, so I do think pursuing it will be worthwhile (no longer needing to port in-progress changes between versions if you didn't make the release cut-off, no need to guard against improper changes to prior versions, or worries about backwards incompatible changes in the non-module code necessitating things like `Mapper2` classes, etc).